doc/developer/design/20230528_managed_clusters.md

```diff
@@ -0,0 +1,349 @@
+
+`ALTER CLUSTER` will support enabling the `MANAGED` option for a cluster to
+convert an unmanaged cluster to a managed cluster:
+
+```sql
+ALTER CLUSTER c SET (MANAGED = true);
+ALTER CLUSTER c SET (MANAGED); -- Shorthand for the above
+```
+
+Conversion is possible if and only if all of the following conditions are met:
+
+  * All replicas of the cluster have the same size `S`.
+  * The replicas are named `r1`, `r2`, ..., `rN`, with no gaps in the numbering.
+    (Zero replicas trivially matches this pattern with `N=0`.)
+
+The `SIZE` option of the cluster will be set to `S` and the
+`REPLICATION FACTOR` will be set to `N`.
+
+Because the conditions for conversion are strict
+
+```sql
+> CREATE CLUSTER c REPLICAS (
+  r1 (SIZE 'small'),
+  r2 (SIZE 'medium')
+);
+> ALTER CLUSTER c SET (MANAGED);
+ERROR: cannot convert cluster "c" to a managed cluster
+DETAIL: Replicas do not all have the same size.
+
+> CREATE CLUSTER c REPLICAS (
+  foo (SIZE 'small'),
+  bar (SIZE 'small'),
+  r3 (SIZE 'small')
+);
+> ALTER CLUSTER c SET (MANAGED);
+ERROR: cannot convert cluster "c" to a managed cluster
+DETAIL: The following replicas do not match required naming pattern `r1`, `r2`, ...
+  foo
+  bar
+
+> CREATE CLUSTER c REPLICAS (
+  r1 (SIZE 'small'),
+  r3 (SIZE 'medium')
+);
+> ALTER CLUSTER c SET (MANAGED);
+ERROR: cannot convert cluster "c" to a managed cluster
+DETAIL: The replicas do not match required naming pattern `r1`, `r2`, .... The
+first missing replica is "r2".
+```
+
+#### Converting an unmanaged cluster to a managed cluster
+
+`ALTER CLUSTER` will support disabling the `MANAGED` option for a cluster to
+convert a managed cluster to an unmanaged cluster:
+
+```sql
+ALTER CLUSTER c SET (MANAGED = false);
+```
+
+There are no constraints on the conversion. All managed clusters can be
+converted to unmanaged clusters.
+
+The conversion preserves the existing replicas of the cluster. Users can then
+create or drop replicas of the cluster using `{CREATE|DROP} CLUSTER REPLICA`.
+
+#### `{CREATE|DROP} CLUSTER REPLICA`
+
+The `CREATE CLUSTER REPLICA` and `DROP CLUSTER REPLICA` statements will be
+restricted to operations on replicas of an unmanaged cluster.
+
+Materialize will produce an error message like the following when attempting
+to manually manipulate replicas of a managed cluster:
+
+```
+> CREATE CLUSTER c SIZE 'small';
+> DROP CLUSTER REPLICA c1.r1;
+ERROR: cannot drop replica of managed cluster "c"
+HINT: Use ALTER CLUSTER to change the cluster's size and replication factor, or
+convert the cluster to an unmanaged cluster by running ALTER CLUSTER c SET (MANAGED = false).
+```
+
+#### `ALTER CLUSTER [REPLICA] ... RENAME TO`
+
+`ALTER CLUSTER` and `ALTER CLUSTER REPLICA` will learn to support renaming
+clusters and cluster replicas, respectively:
+
+```sql
+-- Rename cluster c1 to c2.
+ALTER CLUSTER c1 RENAME TO c2;
+
+-- Rename replica r1 of cluster c1 to r2.
+ALTER CLUSTER REPLICA c1.r1 RENAME TO r2;
+```
```

Comment:
This seems like something we can and should support independently of the managed clusters proposal!

Reply:
Indeed!